modules/nw-ingress-sharding-namespace-labels.adoc

```diff
@@ -0,0 +1,56 @@
+// Module included in the following assemblies:
+//
+// * configuring_ingress_cluster_traffic/configuring-ingress-cluster-traffic-ingress-controller.adoc
+// * ingress-operator.adoc
+
+[id="nw-ingress-sharding-namespace-labels_{context}"]
+= Configuring ingress controller sharding by using namespace labels
+
+Ingress Controller sharding by using namespace labels means that the Ingress
+Controller serves any route in any namespace that is selected by the namespace
+selector.
+
+Ingress Controller sharding is useful when balancing incoming traffic load among
+a set of Ingress Controllers and when isolating traffic to a specific Ingress
+Controller. For example, company A goes to one Ingress Controller and company B
+to another.
+
+.Procedure
+
+. Edit the `router-internal.yaml` file:
++
+----
+# cat router-internal.yaml
+apiVersion: v1
+items:
+- apiVersion: operator.openshift.io/v1
+  kind: IngressController
+  metadata:
+    name: sharded
+    namespace: openshift-ingress-operator
+  spec:
+    domain: <apps-sharded.basedomain.example.net>
+    endpointPublishingStrategy:
+      type: LoadBalancerService
```

Comment:
I don't think we can configure `endpointPublishingStrategy` here, the ingress operator will handle it based on different platforms. Please delete the two lines.

Reply:
The user _can_ specify `endpointPublishingStrategy`, but it is better to omit it, as you suggested, so that the operator will use an appropriate default.

Reply:
Sorry, I was assuming that the procedure was creating a new ingresscontroller, but it uses `oc apply`, which can create a new ingresscontroller _or_ update an existing one.

When creating a new ingresscontroller, `spec.endpointPublishingStrategy` can be specified (or left unspecified, to express a desire for the default value for the platform), and its value will determine the effective endpoint publishing strategy.  If the example is creating a new ingresscontroller, it is better to omit `spec.endpointPublishingStrategy`, as we both said earlier, so that the operator will use the platform default.  (A user who knows better can always add it to the example yaml.)

When updating an existing ingresscontroller, specifying `spec.endpointPublishingStrategy` does not change the effective endpoint publishing strategy (`spec.endpointPublishingStrategy` will have the specified value after the update, but `status.endpointPublishingStrategy`, which shows the effective value, will remain unchanged).  This is also true if `spec.endpointPublishingStrategy` is not specified—after the update, the ingresscontroller will not have `spec.endpointPublishingStrategy` set, but it will still have `status.endpointPublishingStrategy` set.  Having different values in `spec.endpointPublishingStrategy` and in `status.endpointPublishingStrategy` could be confusing.  Having `status.endpointPublishingStrategy` but not `spec.endpointPublishingStrategy` set could be confusing if the user set it when creating the ingresscontroller to override the default, but this scenario is less likely to be confusing, I think, so omitting `spec.endpointPublishingStrategy` from the example makes sense in this case too.

In either case, this procedure should clearly state whether it is creating a new ingresscontroller or updating an existing one (documenting both would be fine too).  For creating a new one, I would suggest using `oc create`, and for updating an existing one, I would suggest using `oc patch`, e.g.:
```shell
oc -n openshift-ingress-operator patch ingresscontroller/sharded --type=merge --patch='{"spec":{"routeSelector":{"matchLabels":{"type":"sharded"}}}}'
```
or
```shell
oc -n openshift-ingress-operator patch ingresscontroller/sharded --type=merge --patch='{"spec":{"nodeSelector":{"matchLabels":{"type":"sharded"}}}}'
```
Sharding can be disabled on an existing ingresscontroller by specifying `null`:
```shell
oc -n openshift-ingress-operator patch ingresscontroller/sharded --type=merge --patch='{"spec":{"routeSelector":null}}'
```
or
```shell
oc -n openshift-ingress-operator patch ingresscontroller/sharded --type=merge --patch='{"spec":{"nodeSelector":null}}'
```